features/mbedtls/targets/TARGET_Silicon_Labs/crypto_aes.c

```diff
@@ -0,0 +1,545 @@
+ *
+ *  http://csrc.nist.gov/encryption/aes/rijndael/Rijndael.pdf
+ *  http://csrc.nist.gov/publications/fips/fips197/fips-197.pdf
+ */
+
+#include "mbedtls/aes.h"
+
+#if defined(MBEDTLS_AES_C)
+#if defined(MBEDTLS_AES_ALT)
+
+#include "em_device.h"
+
+#if defined(CRYPTO_PRESENT)
+
+#include "crypto_management.h"
+#include "em_crypto.h"
+#include "em_core.h"
+#include <string.h>
+
+__STATIC_INLINE void CRYPTO_DataReadUnaligned(volatile uint32_t * reg,
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not, read into temporary buffer and
+     then move to user buffer. */
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
+    CRYPTO_DataRead(reg, temp);
+    memcpy((void*)val, temp, 16);
+  }
+  else
+  {
+    CRYPTO_DataRead(reg, (uint32_t*)val);
+  }
+}
+
+__STATIC_INLINE void CRYPTO_DataWriteUnaligned(volatile uint32_t * reg,
+                                               const uint8_t * val)
+{
+  /* Check data is 32bit aligned, if not move to temporary buffer before
+     writing.*/
+  if ((uint32_t)val & 0x3)
+  {
+    uint32_t temp[4];
+    memcpy(temp, val, 16);
+    CRYPTO_DataWrite(reg, temp);
+  }
+  else
+  {
+    CRYPTO_DataWrite(reg, (uint32_t*)val);
+  }
+}
+
+/*
+ * Initialize AES context
+ */
+void mbedtls_aes_init( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * Clear AES context
+ */
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL ) {
+        return;
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/*
+ * AES key schedule (encryption)
+ */
+int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    if( ctx == NULL || key == NULL ) {
+        return( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+    memcpy(ctx->key, key, keybits/8);
+
+    return 0;
+}
+
+/*
+ * AES key schedule (decryption)
+ */
+int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx,
+                            const unsigned char *key,
+                            unsigned int keybits )
+{
+    CORE_DECLARE_IRQ_STATE;
+
+    if( ctx == NULL || key == NULL ) {
+        return ( MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH );
+    }
+
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+
+    if ( ( 128UL != keybits ) && ( 256UL != keybits ) ) {
+        /* Unsupported key size */
+        return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    ctx->keybits = keybits;
+
+    CRYPTO_TypeDef *device = crypto_management_acquire();
+    device->WAC = 0;
+    device->CTRL = 0;
+
+    CORE_ENTER_CRITICAL();
```

Comment:
The resource we are protecting with critical sections is writing to or reading from 128-bit, 256-bit or 512-bit registers in our accelerator. Since that operation is done with actionable reads/writes from/to a single 32-bit register, they have to be protected against preemption.

The other registers (such as WAC and CTRL) are regular registers, which we get control over by acquiring the device. In the case of preemption, these values will be backed up and put back by the preempting code, so their use is safe.